includes/rtm/impl/vector4f_swizzle.h

```diff
@@ -0,0 +1,193 @@
+    }
+#endif
+
+#if defined(RTM_AVX2_INTRINSICS)
+	template<> RTM_FORCE_INLINE vector4f vector_swizzle_impl<0, 0, 0, 0>(const vector4f& vec)
+    {
+        return _mm_broadcastss_ps(vec);
+    }
+#endif
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Float replicate
+	template<int index>
+	RTM_FORCE_INLINE vector4f vector_replicate_impl(const vector4f& vec)
+	{
+		static_assert(index >= 0 && index <= 3, "Invalid Index");
+		return vector_swizzle_impl<index, index, index, index>(vec);
+	}
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Float shuffle
+	template<int index0, int index1, int index2, int index3>
+	RTM_FORCE_INLINE vector4f vector_shuffle_impl(const vector4f& vec1, const vector4f& vec2)
+	{
+		static_assert(index0 >= 0 && index0 <= 3 && index1 >= 0 && index1 <= 3 && index2 >= 0 && index2 <= 3 && index3 >= 0 && index3 <= 3, "Invalid Index");
+		return _mm_shuffle_ps(vec1, vec2, SHUFFLE_MASK(index0, index1, index2, index3));
+	}
+
+	// Float Shuffle specializations
+	template<> RTM_FORCE_INLINE vector4f vector_shuffle_impl<0, 1, 0, 1>(const vector4f& vec1, const vector4f& vec2)
+    {
+        return _mm_movelh_ps(vec1, vec2);
+
+    }
+	template<> RTM_FORCE_INLINE vector4f vector_shuffle_impl<2, 3, 2, 3>(const vector4f& vec1, const vector4f& vec2)
+    {
+        // Note: movehl copies first from the 2nd argument
+        return _mm_movehl_ps(vec2, vec1);
+    }
+
+#elif defined(RTM_NEON_INTRINSICS)
+    template <int element_index>
+    RTM_FORCE_INLINE vector4f vector_replicate_impl(const vector4f& vec)
+    {
+        return vdupq_n_f32(vgetq_lane_f32(vec, element_index));
+    }
+
+#if defined(__clang__)
+    template <int x, int y, int z, int w>
+    RTM_FORCE_INLINE vector4f vector_swizzle_impl(vector4f vec)
+    {
+        return __builtin_shufflevector(vec, vec, x, y, z, w);
+    }
+
+    template <int x, int y, int z, int w>
+    RTM_FORCE_INLINE vector4f vector_shuffle_impl(vector4f vec1, vector4f vec2)
+    {
+        return __builtin_shufflevector(vec1, vec2, x, y, z + 4, w + 4);
+    }
+#else
+
+    template<int index0, int index1, int index2, int index3>
+    RTM_FORCE_INLINE vector4f vector_shuffle_impl(vector4f vec1, vector4f vec2)
+    {
+        static_assert(index0 <= 3 && index1 <= 3 && index2 <= 3 && index3 <= 3, "Invalid Index");
+
+        static constexpr uint32_t control_element[8] =
+        {
+            0x03020100, // XM_PERMUTE_0X
+            0x07060504, // XM_PERMUTE_0Y
+            0x0B0A0908, // XM_PERMUTE_0Z
+            0x0F0E0D0C, // XM_PERMUTE_0W
+            0x13121110, // XM_PERMUTE_1X
+            0x17161514, // XM_PERMUTE_1Y
+            0x1B1A1918, // XM_PERMUTE_1Z
+            0x1F1E1D1C, // XM_PERMUTE_1W
+        };
+
+        uint8x8x4_t tbl;
+        tbl.val[0] = vget_low_f32(vec1);
```

Comment:
I'm a little concerned with this. As a fallback for non-specialized code paths it might be okay but is the compiler able to identify patterns to use specialized instructions when possible (e.g. zip)? When I last reviewed the assembly for the old vector mix code on ARM, I found that it was sensible. Why change it?